sources/lib/api/gcp/context/_utils.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2016 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Support for getting gcloud credentials. """
+
+import argparse
+import oauth2client.client
+import oauth2client.file
+import oauth2client.tools
+import json
+import os
+
+
+def get_config_dir():
+  config_dir = os.getenv('CLOUDSDK_CONFIG')
+  if config_dir is None:
+    if os.name == 'nt':
+      try:
+        config_dir = os.path.join(os.environ['APPDATA'], 'gcloud')
+      except KeyError:
+        # This should never happen unless someone is really messing with things.
+        drive = os.environ.get('SystemDrive', 'C:')
+        config_dir = os.path.join(drive, '\\gcloud')
+    else:
+      config_dir = os.path.join(os.path.expanduser('~'), '.config/gcloud')
+  return config_dir
+
+
+def get_credentials():
+  """ Get the credentials to use. We try application credentials first, followed by
+      user credentials. If neither of these exist we will do an auth flow provided we are
+      not running inside Docker. The path to the application credentials can be overridden
+      by pointing the GOOGLE_APPLICATION_CREDENTIALS environment variable to some file;
+      the path to the user credentials can be overridden by pointing the CLOUDSDK_CONFIG
+      environment variable to some directory (after which we will look for the file
+      $CLOUDSDK_CONFIG/gcloud/credentials). Unless you have specific reasons for
+      overriding these the defaults should suffice.
+  """
+  try:
+    credentials = oauth2client.client.GoogleCredentials.get_application_default()
+  except Exception as e:
+
+    # Try load user creds from file
+    cred_file = get_config_dir() + '/credentials'
+    if cred_file is not None and os.path.exists(cred_file):
+      with open(cred_file) as f:
```

Comment:
Let me try explain this better: right now you don't have to use gcloud auth or gcloud beta auth anything. You can just go right ahead and use gsutil etc in your %bash cells. That's a pretty nice user experience that I think is worth having.

In order to do this we need user creds accessible to gcloud CLI in the container. So we write these in our node server, but we also write application creds there. Datalab can get by on the application creds so so far there is no need for the code above.

However, if we want to let users map in user creds from outside for gcloud CLI to use, then we need to let them customize CLOUDSDK_CONFIG to point to the credentials file in the mapped volume. And it makes sense for us to make use of these in Datalab too, because if we don't, the node server is going to think there are no creds and will cause re-authentication to occur, which kind of defeats the purpose of having the creds mapped in from outside. Hence we support reading these creds here as an alternative for Datalab.

There are other options, if we want to support gcloud CLI and Datalab using credentials mapped in from outside:

- we could require the user to set both GCLOUDSDK_CONFIG and GOOGLE_APPLICATION_CREDENTIALS; Datalab will use the latter and gcloud the former. But we can make things a bit easier on the user by having these few lines here;
- we could have the node server detect if there are user credentials but no application credentials (so the user mapped in the user creds from outside) and have it read those and generate an application_credentials file instead of doing so by initiating auth; that would allow the Python code to rely just on application credentials while still achieving the same effect. 

If we were to change this I'd argue going for the latter approach (have node create app credentials if there are none but there are user credentials).